[WIP] LayoutPanel: Update to use ToolsPanel and ToolsPanelItems

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

🚧 🚧 🚧 🚧 This is still a draft, it isn't _quite_ ready for testing yet 🚧 🚧 🚧 🚧 

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes https://github.com/WordPress/gutenberg/issues/44560

Update the `LayoutPanel` which appears in the inspector controls to use the `ToolsPanel` so that controls can be conditionally displayed and reset.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As additional controls are added to the layout panel, some of which might be used infrequently, it will be helpful if we can hide some of them by default. A good example is sticky position support (as raised in https://github.com/WordPress/gutenberg/issues/30121 and being worked on in https://github.com/WordPress/gutenberg/pull/44723), which will likely be a useful feature that will only be used in particular circumstances.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a `layout` group to the inspector controls
* Wrap each of the layout controls in a `ToolsPanelItem`

## To-do

* [ ] Tweak broken styling / subtle styling issues
* [ ] Wrap layout switcher in a `ToolsPanelItem`
* [ ] Work out what the correct defaults should be for each block that opts-in to the layout block support
* [ ] Double-check that it's working correctly for each block that uses layout, and update the testing instructions to reflect this

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

TBD

## Screenshots or screencast <!-- if applicable -->

| Constrained layout | Flex layout |
| --- | --- |
| ![2022-11-17 14 13 27](https://user-images.githubusercontent.com/14988353/202358612-d5aa9958-797a-4a3f-a4f9-5b70a1088af6.gif) | ![2022-11-17 15 49 54](https://user-images.githubusercontent.com/14988353/202359241-6e4ccbf2-000d-4eba-a5c3-5f0d5f4e9c51.gif) |

andrewserong committed:
LayoutPanel: Update to use ToolsPanel and ToolsPanelItems

andrewserong commented:
One the potentially challenging design issues for this one: The Justification and Orientation controls bump up against each other, as the Justification controls extend slightly past the gutter that's meant to exist between these two controls:

<img width="282" alt="image" src="https://user-images.githubusercontent.com/14988353/202975047-53df91fc-4b49-4c0b-97ad-9ea0f0e58cd1.png">

andrewserong committed:
Wrap layout type switcher in ToolsPanel, fix styling of content and wide controls

andrewserong commented:
@jasmussen and @jameskoster 👋 just a heads-up that I'm working on this one now. It seems to be going pretty well so far, but there are a couple of styling issues to address. The main issue is currently with the spacing between the Justification and Orientation controls — they _just_ manage to fit in next to each other, but it's a bit of a squeeze.

There's still a bit more to work on for this PR (I haven't tested it out across blocks, we'll need to tweak defaults, etc), but just thought I'd let you know that work is under way!

jameskoster commented:
Thanks for working on this.

One initial observation: I don't think "Content size" should be a dedicated control. It's intrinsically linked to the "Content layout" option, and should be presented automatically when the latter is set to true.

---

When I toggle the content layout option to true, the Justification control appears, is that intended? I couldn't really work out why that happened and my expectation is that it would remain toggled off until explicitly enabled via the tools menu.

---

As you mentioned, when there are four options the Justification and Orientation controls are breaking the grid:

<img width="476" alt="Screenshot 2022-11-21 at 10 31 48" src="https://user-images.githubusercontent.com/846565/203028238-df51925d-ffd9-4b12-b254-f12148a94717.png">

Probably best to place them on separate rows for now because we'll likely need to revisit these controls – they do not allow Stacks to be have right/center justification with space-between spacing.

andrewserong committed:
Prevent constrained justification from showing by default

andrewserong committed:
Display justification and orientation on separate lines

andrewserong committed:
Try moving content layout down to the layout type, combine with contentSize and wideSize

andrewserong commented:
Thanks for the testing and feedback @jameskoster! I've made a few updates:

* Fixed issue where justification was displaying by default in the Group block when the content width toggle was switched on
* Updated the UI to display the justification and orientation controls on separate lines, and tweaked the spacing for consistency
* Refactored the content width toggle so that it's combined in the same `ToolsPanelItem` as the content and wide size controls. I agree, it feels better keeping all of those controls combined under the one item. Now that they're combined into the same files, it should be a little easier to pull them apart again if further down the track we decide we'd like to add more fine-grained control over resetting these items.

I had to move things around a fair bit while doing the updates, so I very likely might have broken something in the process. I'll do some more testing and follow-up tomorrow, so no rush on taking another look at this just yet 🙂

jasmussen commented:
I realize I'm a bit late to this, but I think showing these two on a single row may be worth breaking the 2 column grid for:

<img width="314" alt="image" src="https://user-images.githubusercontent.com/1204802/203267576-c8c3945b-83d1-4cd1-afc8-8069c003798e.png">

I think there may be a separate exploration worth having, of making the orientation + justification controls count as a single control 🤔

jameskoster commented:
> I think there may be a separate exploration worth having, of making the orientation + justification controls count as a single control 🤔

Yeah I think that makes sense. But as with #44723 if we're going to consider the full layout scope, we need to consider things like container height and width. It may be that the entire layout panel is a single control, like in Figma which I personally find to be very intuitive. We can potentially find inspiration there:


https://user-images.githubusercontent.com/846565/203312971-b2a1602a-44e2-4dab-9150-267239b91d89.mp4

* Adding layout to a Group inherently transforms it into a Stack
* Orientation, alignment/justification, and spacing mode are all linked
* The `–` icon removes the layout altogether and reverts to Group.

jasmussen commented:
Intriguing! Some good bits there worth noodling on. Some things would need refinement, but at a more high level, I wonder if it would make sense to think of this in terms of such a boolan "having layout" toggle, with layout being _flex_ and no layout being _block_ 🤔

jameskoster commented:
Yep that's exactly what I was thinking :)

jasmussen commented:
It would be nice to still have a single "Layout" panel, though, so I wonder if there was a way to have these `has-layout` tools work in the existing environment 🤔

jameskoster commented:
Perhaps the options here could be revised to represent flex/flow:

<img width="234" alt="Screenshot 2022-11-22 at 13 47 41" src="https://user-images.githubusercontent.com/846565/203330057-d44a8b61-5fbf-456f-b98d-a6009db0d9c1.png">

And depending on which option you select determines which controls are available in the Layout panel. That would also solve the awkward duplication we have currently between this variation switcher and the orientation control.

andrewserong commented:
Love the ideation, folks! I'm very happy seeing all the attention to exploring our Layout panel options 😀
Hope you don't mind, but just jotting down some immediate thoughts:

> Intriguing! Some good bits there worth noodling on. 

Very much so. I suppose at the moment, my main focus is to try to figure out what makes for a good logical next step in this PR, where the UX feels decently improved, but without overhauling how we conceptualise layout just yet. Still, it's super valuable taking a look at the direction we'd like to go in, to confidence check that we're taking steps in the right direction 👍

At the moment, the layout concept is more of an `is-a` logic rather than a `has-a` logic, in that the default layout type is flow, which defines spacing as using top-margins, so it isn't quite a blank slate. Separately, I was wondering if eventually we might have a case for a `null` layout type that doesn't even apply a particular spacing type? Also worth considering for how we handle the UI is that we have `constrained` as a discrete layout type, and the current layout types are [documented in the handbook](https://developer.wordpress.org/block-editor/explanations/architecture/styles/#available-layout-types), so whichever changes we wind up making, we'll need to take care that they feel like a logical progression for folks already used to how they work.

So, I think I lean a bit more toward thinking of the different layout types as adjacent rather than additive, if that makes sense? E.g. we can switch between layouts, with flow as default, rather than thinking about adding a layout per se — but this might just be because I've been working closely with it for so long, so I might have a familiarity bias here 🙂

One technical note is that I don't think we have a way for the plus icon in the ToolsPanel menu to do anything other than display the current menu items available, so having that button perform an action would be a departure from the other instances in the sidebar and could be a challenge figuring out the right accessibility for it.

> Orientation, alignment/justification, and spacing mode are all linked
> The – icon removes the layout altogether and reverts to Group.

If we group all the controls together, does that then defeat the purpose of switching over to the ToolsPanel? I think the panel is most useful for granular control over the display and resetting of separate settings. The change in this PR is probably going to be most valuable if we have seldom used controls that we'd like users to be able to reach for only when needed (the sticky position support being the more pressing use case).

> And depending on which option you select determines which controls are available in the Layout panel. That would also solve the awkward duplication we have currently between this variation switcher and the orientation control.

One of the other things to keep in mind with this PR is that the changes affect all blocks that opt in to layout, and not just the Group block. So while there's duplication of the control for Row and Stack, there is no duplication in Buttons and Social Icons blocks.

---

Anyhow, thanks again for all the explorations! If we're veering off track from the current task of the conversion to ToolsPanel, I'm happy for us to continue the discussion in separate issues if need be.

For the moment, what do you think the best next step is for this PR design-wise? Is the main thing to address what we do about justification and orientation controls?

jameskoster commented:
The layout concept has many threads, and the current implementation is not perfect – particularly how the controls switch based on whether the block is flow or flex. 

So my instinct is that this needs a bit more design consideration before making any changes to the layout panel in the short term, for the reasons above, and because the migration seems to require some tangential effort – grouping certain controls, presentation of options like "Inner blocks use content width", etc.

andrewserong commented:
> So my instinct is that this needs a bit more design consideration before making any changes to the layout panel in the short term, for the reasons above, and because the migration seems to require some tangential effort – grouping certain controls, presentation of options like "Inner blocks use content width", etc.

Understood, thanks for clarifying! I'll pause work on this PR for now, then.

In terms of the sticky position support over in https://github.com/WordPress/gutenberg/pull/44723, do you think it's still worth pursuing for now, or does that also need more design-time to figure out how it'll work inside the panel? From my perspective, sticky position will be used a bit too infrequently for it to be displayed prominently in the existing panel without the ToolsPanel for it to be hidden by default, but happy to go with your or @jasmussen's lead 🙂

tellthemachines commented on packages/block-editor/src/layouts/flow.js:
```diff
@@ -2,19 +2,85 @@
+		// Only show the inherit toggle if it's supported,
+		// a default theme layout is set (e.g. one that provides `contentSize` and/or `wideSize` values),
+		// and either the default / flow or the constrained layout type is in use, as the toggle switches from one to the other.
+		const showInheritToggle = !! (
```
Could we keep this logic in the parent layout panel, so as not to duplicate it across two layout types?

andrewserong replied:
> Could we keep this logic in the parent layout panel, so as not to duplicate it across two layout types?

Good question! I couldn't work out a way to combine this toggle with the content size controls without moving it down to the layout type level. If we keep them separate (as you suggest in your other comment), then we can happily move it back to the parent.

Let me know if you can think of a better way to go about it, though! (I'll hold off on implementing further changes until we hear more from designers)

tellthemachines reviewed (commented):
Thanks for working on this! Just catching up on the discussion now and adding a few thoughts:

* When adding a block with "constrained" type I was hoping that the controls for custom content and wide size would no longer display by default. In the multiple discussions around the non-intuitiveness of layout controls 😅 one point that came up a bit was that it generally makes more sense for users to be able to set a global default than a custom size for each block. Given these controls may not get much use, it would be nice to separate them logically from the "use content width" toggle and have them hidden by default.
* It feels a bit weird to have "allow to wrap" control display by default for the horizontal flex blocks, whereas for vertical it's "orientation". Perhaps we could display orientation and justification by default for both?

>  I wonder if it would make sense to think of this in terms of such a boolan "having layout" toggle, with layout being flex and no layout being block

I'm not sure this will be possible at this point. We already have two non-flex layout types ("flow" and "constrained"), "position" will be another and in the future we also want to implement "grid".

> the current implementation is not perfect – particularly how the controls switch based on whether the block is flow or flex

I'm inclined to consider this a feature 😄 because they are different layout types with different features, so it makes sense that they have different controls. Having the same-looking control potentially have different effects for different layout types would be quite confusing. 

> So my instinct is that this needs a bit more design consideration before making any changes to the layout panel in the short term

My biggest concern here is that we're considering this change as an essential part of the work towards stabilising the layout support, which we should do ASAP because the feature is starting to see some real-world use, and the more extenders come to depend on it, the less flexibility we have to make changes going forward. 

Could we perhaps turn this around and _not_ consider the migration to ToolsPanel essential, as it is mostly a cosmetic rearrangement of the controls, and focus instead on stabilising the API for extenders?

jameskoster commented:
> do you think it's still worth pursuing for now, or does that also need more design-time to figure out how it'll work inside the panel?

That one needs a little more work. It came up in a conversation yesterday with Matias and Joen, and consensus was to take another look at fixed & sticky together to make sure we're not painting ourselves into a corner with the design. It may still be that we implement Sticky on its own first. Hopefully we'll have an update soon :) 

> I'm inclined to consider this a feature 😄 because they are different layout types with different features, so it makes sense that they have different controls.

Sorry I was referring to the UX. I understand the need for different sets of controls, but we need to present them in ways that feel intuitive and the current experience is a bit lacklustre in that regard. For instance it's not really clear why switching from Group to Stack results in totally different Layout controls. Or that 'Orientation' and block type (Row / Stack) are the same thing presented two different ways.

andrewserong commented:
Thanks for the feedback and updates, folks!

I think it was a useful exploration this PR, in digging up these ideas and insights into the controls. For now, I think let's park this PR,particularly with the discussion over in https://github.com/WordPress/gutenberg/discussions/46032 that looks at other options for where the sticky/fixed position control could go 👍

> Could we perhaps turn this around and not consider the migration to ToolsPanel essential, as it is mostly a cosmetic rearrangement of the controls, and focus instead on stabilising the API for extenders?

Treating the ToolsPanel migration as non-essential sounds like a good idea to me, as that'll free up other more impactful work if we can workaround the absence of the ToolsPanel like in https://github.com/WordPress/gutenberg/discussions/46032

andrewserong commented:
Since we're pursuing a different direction for the Position UI now that is decoupled from Layout, and this PR is a little old, I'll close out this experiment. We can always open a new PR and borrow ideas from here if and when we'd like to revisit using the ToolsPanel for Layout.
